Split the `Edges` iterator.

nnethercote opened the pull request:
Some nice performance wins here, mostly on the `wg-grammar` benchmark.

r? @lcnr

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b49a7349f77581289529cf95495e4b9a33b704ad with merge 449a9f79d82ced53005ec8aac40306afa5af4cbc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b49a7349f77581289529cf95495e4b9a33b704ad","merge_sha":"449a9f79d82ced53005ec8aac40306afa5af4cbc"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13537988113/job/37837884561)
Build commit: 449a9f79d82ced53005ec8aac40306afa5af4cbc (`449a9f79d82ced53005ec8aac40306afa5af4cbc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13537988113/job/37837884561"},"merge_sha":"449a9f79d82ced53005ec8aac40306afa5af4cbc"} -->

rust-timer commented:
Finished benchmarking commit (449a9f79d82ced53005ec8aac40306afa5af4cbc): [comparison URL](https://perf.rust-lang.org/compare.html?start=a46c755c6ad2c9543a836a1ca134178b2345acc2&end=449a9f79d82ced53005ec8aac40306afa5af4cbc&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a46c755c6ad2c9543a836a1ca134178b2345acc2&end=449a9f79d82ced53005ec8aac40306afa5af4cbc&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.8% | [-5.3%, -4.5%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a46c755c6ad2c9543a836a1ca134178b2345acc2&end=449a9f79d82ced53005ec8aac40306afa5af4cbc&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=a46c755c6ad2c9543a836a1ca134178b2345acc2&end=449a9f79d82ced53005ec8aac40306afa5af4cbc&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=a46c755c6ad2c9543a836a1ca134178b2345acc2&end=449a9f79d82ced53005ec8aac40306afa5af4cbc&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 771.663s -> 772.21s (0.07%)
**Artifact size**: 361.93 MiB -> 361.97 MiB (0.01%)

nnethercote committed:
Adjust `ConstraintGraphDirection` methods.

This facilitates the next commit.

nnethercote committed:
Split the `Edges` iterator.

The `Edges` iterator returns `OutlivesConstraint` elements, which are 72
bytes. This is big enough to affect performance. Return
`&OutlivesConstraint` would be better. However, each `Edges` iterator is
really one of two different iterators. The "from graph" case does a
graph traversal and could return `&OutlivesConstraint`. But the "from
static" case just does a `0..n` iteration and constructs a new
`OutlivesConstraint` from that, so it can't return a reference.

This commit splits `Edges into `EdgesFromGraph` and `EdgesFromStatic`,
which allows them to have different return types. This is a perf win for
the `wg-grammar` benchmark.

nnethercote committed:
Optimize `Trace`.

`Trace::FromOutlivesConstraint` contains an `OutlivesConstraint`, which
is 72 bytes. Lots of these are created but never used.

This commit splits `Trace::FromOutlivesConstraint` into three new
variants: `FromVanilla`, `FromStatic`, `FromMember`. Each of these
contains just enough data to construct an `OutlivesConstraint`, if
necessary. This reduces the size of `Trace` from 72 bytes to 16 bytes.
All this avoids some memory traffic.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 48a95d1b23aca50c58082be7de912206751567b5 with merge 920033be46e1f50b03632d1215206f9e22801e8d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"48a95d1b23aca50c58082be7de912206751567b5","merge_sha":"920033be46e1f50b03632d1215206f9e22801e8d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13556591018/job/37895863549)
Build commit: 920033be46e1f50b03632d1215206f9e22801e8d (`920033be46e1f50b03632d1215206f9e22801e8d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13556591018/job/37895863549"},"merge_sha":"920033be46e1f50b03632d1215206f9e22801e8d"} -->

rust-timer commented:
Queued 920033be46e1f50b03632d1215206f9e22801e8d with parent 00f245915b0c7839d42c26f9628220c4f1b93bf6, future [comparison URL](https://perf.rust-lang.org/compare.html?start=00f245915b0c7839d42c26f9628220c4f1b93bf6&end=920033be46e1f50b03632d1215206f9e22801e8d).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (920033be46e1f50b03632d1215206f9e22801e8d): [comparison URL](https://perf.rust-lang.org/compare.html?start=00f245915b0c7839d42c26f9628220c4f1b93bf6&end=920033be46e1f50b03632d1215206f9e22801e8d&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=00f245915b0c7839d42c26f9628220c4f1b93bf6&end=920033be46e1f50b03632d1215206f9e22801e8d&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.1% | [-5.6%, -4.8%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=00f245915b0c7839d42c26f9628220c4f1b93bf6&end=920033be46e1f50b03632d1215206f9e22801e8d&stat=max-rss)
<details>
<summary>Results (primary -2.7%, secondary -2.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.7% | [-3.1%, -1.9%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.9%, -2.2%] | 2     |
| All ❌✅ (primary)                 | -2.7% | [-3.1%, -1.9%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=00f245915b0c7839d42c26f9628220c4f1b93bf6&end=920033be46e1f50b03632d1215206f9e22801e8d&stat=cycles:u)
<details>
<summary>Results (secondary -4.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.8% | [-6.4%, -2.3%] | 8     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=00f245915b0c7839d42c26f9628220c4f1b93bf6&end=920033be46e1f50b03632d1215206f9e22801e8d&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 770.979s -> 770.86s (-0.02%)
**Artifact size**: 361.96 MiB -> 361.97 MiB (0.00%)

lcnr reviewed (commented):
not too big a fan of `EdgesVanilla` as a name :thinking: maybe just `Edges` and `EdgesFromStatic` iterator :thinking:  

r=me after wards

nnethercote commented:
Ok. I originally had `EdgesNormal` but that was confusing because `Normal` is already used for the graphs (as the opposite of `Reverse`).

nnethercote commented:
After discussion on Zulip with @lcnr I went with `EdgesFromGraph` and `EdgesFromStatic`.

@bors r+

bors commented:
:pushpin: Commit 9a0513ad7134f75d51b16ac359e41f52f9dda463 has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote 9a0513ad7134f75d51b16ac359e41f52f9dda463 -->
<!-- homu: {"type":"Approved","sha":"9a0513ad7134f75d51b16ac359e41f52f9dda463","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/137703#issuecomment-2687678250) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

bors commented:
:hourglass: Testing commit 9a0513ad7134f75d51b16ac359e41f52f9dda463 with merge 385970f0c1fd0c09bac426b02f38300c0b1ba9a2...
<!-- homu: {"type":"BuildStarted","head_sha":"9a0513ad7134f75d51b16ac359e41f52f9dda463","merge_sha":"385970f0c1fd0c09bac426b02f38300c0b1ba9a2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13749931443/job/38452670569)
Approved by: nnethercote
Pushing 385970f0c1fd0c09bac426b02f38300c0b1ba9a2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nnethercote","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13749931443/job/38452670569"},"merge_sha":"385970f0c1fd0c09bac426b02f38300c0b1ba9a2"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13751349564/job/38452696670) [(plain)](https://github.com/rust-lang/rust/commit/385970f0c1fd0c09bac426b02f38300c0b1ba9a2/checks/38452696670/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

gh pr comment ${HEAD_PR} -F output.log
shell: /usr/bin/bash -e {0}
##[endgroup]
fatal: ambiguous argument 'HEAD^1': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
##[error]Process completed with exit code 128.
```

</details>

rust-timer commented:
Finished benchmarking commit (385970f0c1fd0c09bac426b02f38300c0b1ba9a2): [comparison URL](https://perf.rust-lang.org/compare.html?start=ed897d5f85d61d56a7d52b180e34c8ca62b59e30&end=385970f0c1fd0c09bac426b02f38300c0b1ba9a2&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ed897d5f85d61d56a7d52b180e34c8ca62b59e30&end=385970f0c1fd0c09bac426b02f38300c0b1ba9a2&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.1% | [-5.6%, -4.8%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ed897d5f85d61d56a7d52b180e34c8ca62b59e30&end=385970f0c1fd0c09bac426b02f38300c0b1ba9a2&stat=max-rss)
<details>
<summary>Results (primary -0.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-0.8%, -0.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-0.8%, -0.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ed897d5f85d61d56a7d52b180e34c8ca62b59e30&end=385970f0c1fd0c09bac426b02f38300c0b1ba9a2&stat=cycles:u)
<details>
<summary>Results (secondary -5.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.2% | [-5.9%, -4.7%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ed897d5f85d61d56a7d52b180e34c8ca62b59e30&end=385970f0c1fd0c09bac426b02f38300c0b1ba9a2&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 768.783s -> 769.881s (0.14%)
**Artifact size**: 361.99 MiB -> 361.99 MiB (0.00%)
